protocols/kad/src/behaviour.rs

```diff
@@ -1449,6 +1449,46 @@ where
         self.connected_peers.insert(peer.clone());
     }
 
+    fn inject_address_change(
+        &mut self,
+        peer: &PeerId,
+        _: &ConnectionId,
+        old: &ConnectedPoint,
+        new: &ConnectedPoint
+    ) {
+        let (old, new) = (old.get_remote_address(), new.get_remote_address());
+
+        // Update routing table.
+        if let Some(addrs) = self.kbuckets.entry(&kbucket::Key::new(peer.clone())).value() {
+            if addrs.replace(old, new) {
+                debug!("Address '{}' replaced with '{}' for peer '{}'.", old, new, peer);
+            } else {
+                debug!(
+                    "Address '{}' not replaced with '{}' for peer '{}' as old address wasn't \
+                     present.",
+                    old, new, peer,
+                );
+            }
+        } else {
+            debug!(
+                "Address '{}' not replaced with '{}' for peer '{}' as peer is not present in the \
+                 routing table.",
+                old, new, peer,
+            );
+        }
+
+        // Update query address cache.
+        for query in self.queries.iter_mut() {
```

Comment:
I don't think there is a point in doing this, but I may be mistaken. The query address "cache" is used to make sure when a query wants to dial a certain address  that often may not (yet) be in the routing table, that address is returned by `addresses_of_peer`. Addresses in the routing table are always returned from `addresses_of_peer`. So when the new address is successfully put in the routing table, it will be considered for dialing. Please correct me if I'm wrong.

Reply:
> Given two connected nodes: local node A and remote node B. Say node B is not in node A's routing table. Additionally node B is part of the QueryInner::addresses list of an ongoing query on node A. Say Node B triggers an address change and then disconnects. Later on the earlier mentioned query on node A would like to connect to node B. Without replacing the address in the QueryInner::addresses set node A would attempt to dial the old and not the new address.

Fair enough, I guess given the right timing, this situation can occur. I still wonder though whether this is worth this special attention and effort. Maybe there is no good reason, but changing the addresses that a query discovered like this seems strange to me. At least it adds a subtle twist to reasoning about which addresses a query uses. The other aspect is that, depending on how often addresses change like this (certainly increasingly often the more peers you are connected to), iterating through all discovered addresses of that peer in all currently ongoing queries seems like a search for a needle in a haystack, if there is substantial traffic. Neither of these points is a good technical reason against doing this, I just had to express that I feel a bit uncomfortable about it and why that is. I guess the decision is yours, ultimately.

Reply:
> At least it adds a subtle twist to reasoning about which addresses a query uses.

That is a good point.

Still I would prefer to uphold correctness and revisit the decision once we have a transport protocol that supports address changes and the given logic proves to be a performance problem. I have included a note in https://github.com/libp2p/rust-libp2p/pull/1649/commits/ad81fd6146853d9fa08148a70705891317eaaf29 summarizing this discussion.